common/thorhelper/thorsoapcall.cpp

```diff
@@ -2554,7 +2560,54 @@ class CWSCAsyncFor : implements IWSCAsyncFor, public CInterface, public CAsyncFo
                         if (proto == PersistentProtocol::ProtoTLS)
                         {
 #ifdef _USE_OPENSSL
-                            Owned<ISecureSocket> ssock = master->createSecureSocket(socket.getClear(), connUrl.host);
+                            if (useProxy && strieq(connUrl.method.str(), "http"))
+                            {
+                                StringBuffer urlHost;
+                                if (streq(url.host.str(), "."))
+                                    urlHost.append(GetCachedHostName());
+                                else
+                                    urlHost.append(url.host.str());
+                                StringBuffer proxyText;
+                                proxyText.appendf("CONNECT %s:%d HTTP/1.1\r\n", urlHost.str(), url.port);
+                                proxyText.append("Proxy-Connection: Keep-Alive\r\n\r\n");
+                                socket->write(proxyText.str(),proxyText.length());
+
+                                checkTimeLimitExceeded(&remainingMS);
+                                unsigned proxyTimeout = MIN(remainingMS, master->timeoutMS);
+                                char proxyResponse[MAX_HTTP_HEADERSIZE + 1];
+                                unsigned proxyRemaining = proxyTimeout;
+                                CTimeMon tm(proxyTimeout);
+                                unsigned totalRead = 0;
+                                unsigned bytesRead = 0;
+                                bool sockClosed = false;
+                                while (!sockClosed && !tm.timedout(&proxyRemaining))
+                                {
+                                    unsigned maxPoss = MAX_HTTP_HEADERSIZE - totalRead;
+                                    sockClosed = readtmsAllowClose(socket, &proxyResponse[totalRead], 1, maxPoss, bytesRead, proxyRemaining);
+                                    totalRead += bytesRead;
+                                    proxyResponse[totalRead] = 0;
```

Comment:
minor: can this move to after the while loop?

Reply:
no, because we strstr(proxyResponse) in the loop and strstr() needs a null-terminated string.